src/base/FoamProblem.C

```diff
@@ -67,39 +82,40 @@ void
 BuoyantFoamProblem::syncSolutions(Direction dir)
 {
   auto & mesh = static_cast<FoamMesh &>(this->mesh());
-  auto serial = mesh.isSerial();
 
   if (dir == ExternalProblem::Direction::FROM_EXTERNAL_APP)
   {
-
-    std::vector<Real> foamT;
-    foamT.reserve(mesh.nNodes());
+    // Vector to hold the temperature on the elements in every subdomain
+    // Not sure if we can pre-allocate this - we need the number of elements
+    // in the subdomain owned by the current rank. We count this in a loop
+    // later.
+    std::vector<Real> foam_vol_t;
 
     auto subdomains = mesh.getSubdomainList();
-    // First we get all the temperature data for every sobdomain
+    // The number of elements in each subdomain of the mesh
+    // Allocate an extra element as we'll accumulate these counts later
     std::vector<size_t> patch_counts(subdomains.size() + 1, 0);
+    for (auto i = 0U; i < subdomains.size(); ++i)
     {
-      int i = 0;
-      for (auto const & subdom : subdomains)
-      {
-        patch_counts[i++] = _app.append_patchT(subdom, foamT);
-      }
+      patch_counts[i] = _app.append_patch_face_T(subdomains[i], foam_vol_t);
     }
     std::exclusive_scan(patch_counts.begin(), patch_counts.end(), patch_counts.begin(), 0);
+
+    // The offsets for each rank into the MOOSE mesh's elements
+    auto rank_offsets = find_mpi_rank_offsets(patch_counts.back(), MPI_COMM_WORLD);
```

Comment:
I think(?) this could be moved to the mesh set-up, I don't think the counts change.